flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java

```diff
@@ -0,0 +1,338 @@
+                        ? InitialPosition.TRIM_HORIZON
+                        : InitialPosition.valueOf(
+                                consumerConfig
+                                        .getOrDefault(
+                                                STREAM_INITIAL_POSITION,
+                                                DEFAULT_STREAM_INITIAL_POSITION)
+                                        .toString());
+        StartingPosition startingPosition;
+        switch (initialPositionFromConfig) {
+            case LATEST:
+                // If LATEST is requested, we still set the starting position to the time of
+                // startup. This way, the job starts reading from a deterministic timestamp
+                // (i.e. time of job submission), even if it enters a restart loop immediately
+                // after submission.
+                startingPosition = StartingPosition.fromTimestamp(Instant.now());
+                break;
+            case AT_TIMESTAMP:
+                startingPosition =
+                        StartingPosition.fromTimestamp(
+                                parseStreamTimestampStartingPosition(consumerConfig).toInstant());
+                break;
+            case TRIM_HORIZON:
+            default:
+                startingPosition = StartingPosition.fromStart();
+        }
+
+        List<KinesisShardSplit> splits = new ArrayList<>();
+        for (Shard shard : shards) {
+            splits.add(new KinesisShardSplit(streamArn, shard.shardId(), startingPosition));
+        }
+
+        return splits;
+    }
+
+    /**
+     * This method assigns a given set of Kinesis splits to the readers currently registered on the
+     * cluster. This assignment is done via a side-effect on the {@link SplitEnumeratorContext}
+     * object.
+     *
+     * @param discoveredSplits list of discovered splits
+     * @param t throwable thrown when discovering splits. Will be null if no throwable thrown.
+     */
+    private void assignSplits(List<KinesisShardSplit> discoveredSplits, Throwable t) {
+        if (t != null) {
+            throw new KinesisStreamsSourceException("Failed to list shards.", t);
+        }
+
+        if (context.registeredReaders().isEmpty()) {
+            LOG.info("No registered readers, skipping assignment of discovered splits.");
+            unassignedSplits.addAll(discoveredSplits);
+            return;
+        }
+
+        Map<Integer, List<KinesisShardSplit>> newSplitAssignments = new HashMap<>();
+        for (KinesisShardSplit split : unassignedSplits) {
+            assignSplitToSubtask(split, newSplitAssignments);
+        }
+        unassignedSplits.clear();
+        for (KinesisShardSplit split : discoveredSplits) {
+            assignSplitToSubtask(split, newSplitAssignments);
+        }
+
+        updateLastSeenShardId(discoveredSplits);
+        updateSplitAssignment(newSplitAssignments);
+        context.assignSplits(new SplitsAssignment<>(newSplitAssignments));
```

Comment:
Great callout. 

So I have checked that should be effectively synchronous across components due to our RPC setup.


1. **On JM:** assignSplit() runs in the coordinatorThread. [code here](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java#L323). (This is the same thread that handles checkpointCoordinator() that triggers checkpoint on SourceCoordinator). This means the order of calls will be maintained here. 
2. In both cases, these make RPC calls to the TM. [assignSplit() calls `sendOperatorEventToTask()`](https://github.com/apache/flink/blob/2c25292272306a6eb094be40be0a0fa5fc3cf644/flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutorGateway.java#L284) and `checkpointCoordinator()` [calls `triggerCheckpoint()`](https://github.com/apache/flink/blob/2c25292272306a6eb094be40be0a0fa5fc3cf644/flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutorGateway.java#L141). 
3. Since RPC calls are FIFO, we can be assured that the `assignSplit()` will be always processed before `checkpointCoordinator()`

Reply:
Thanks for the deep dive @hlteoh37 !